app/src/main/java/com/example/trainingintive/MyApplication.kt

```diff
@@ -1,10 +1,19 @@
 package com.example.trainingintive
 
 import android.app.Application
+import com.example.trainingintive.data.UserSession
 import com.example.trainingintive.di.AppComponent
 import com.example.trainingintive.di.DaggerAppComponent
+import com.firebase.ui.auth.AuthUI
+import com.google.firebase.auth.FirebaseAuth
 
-class MyApplication : Application() {
+class MyApplication : Application(), UserSession {
 
-    val appComponent: AppComponent by lazy { DaggerAppComponent.factory().create(applicationContext) }
+    val appComponent: AppComponent by lazy { DaggerAppComponent.factory().create(this) }
+
+    override fun logout() {
+        AuthUI.getInstance().signOut(applicationContext)
```

Comment:
I think that the previous version with such logic in the repository was better. The data layer has access to the system so it's nothing wrong with application context in the data layer repository when it is required. Application same as Activities should contain as little logic as possible.